### ISSUE TITLE
Idea: useSingleton

### ISSUE THREAD

**Issue by antfu** [vueuse/vueuse#423]
An SSR-friendly singleton / lazy initialization implementation.

When you call the function, it will use `inject()` to reuse the previous instance. If it's not found, initialize one and provide it to the app root. This could be useful to implement event pools and free up the event registration for each call/each component, for example [`onKeyStoke`](https://vueuse.org/core/onkeystroke/) and #422 without the need to do the global registration like `app.use(xx)`.

Usage example

```ts
const useHey = useAppSingleton(()=>{
  console.log('Hi')
  return ref(1)
})
```

```ts
const a = useHey() // console: 'Hi'
const b = useHey() // no console output, reused
const same = a === b // true
```

## Implementaion

Waiting for the [`effectScope` API](https://github.com/vuejs/rfcs/pull/212) to be shipped, otherwise, it only works for the effect-free factories.

### App based

```ts
function useAppSingleton<T>(factory: () => T, key: string | Symbol = Symbol()): () => T {
  return () => {
    let injected = inject<T>(key)
      if (!injected) {
      const vm = getCurrentInstance()
      const app = vm?.appContext.app
      // TODO: effectScope
      injected = factory()
      app?.provide(key, injected)
    }
    return injected
  }
}
```

### General (not SSR friendly)

```ts
function useSingleton<T>(factory: () => T, lazy = true): () => T {
  let data: T | undefined
  if (!lazy)
    // TODO: effectScope
    data = factory()
  return () => {
    if (!data)
      // TODO: effectScope
      data = factory()
    return data
  }
}
```

**Comment by lstoeferle** [comment]
Will it work with Vue2 also?

**Comment by antfu** [comment]
> Will it work with Vue2 also?

Yeah, I think it could be when effectScope get backported to Vue 2. Which I think could take a while, would focus on Vue 3.1 first.

**Comment by davay42** [comment]
@antfu May be it's related to my recent error after updating a vitepress project from vueuse/core 4.8.2 to 4.8.3

```
 > node_modules/.pnpm/@vueuse+core@4.8.3_vue@3.0.11/node_modules/@vueuse/core/dist/index.esm.js:4:39: error: Could not resolve "@antfu/utils" (mark it as external to exclude it from the bundle)
    4 │ import { createSingletonPromise } from '@antfu/utils'
```

**Comment by antfu** [comment]
@davay42 should be fixed in 4.8.4, thanks! But they are actually not related to this issue.

**Comment by husayt** [comment]
@antfu now that Effectscope api is out, this might be a really useful addition to the library

```
export function useSingleton<T>( key: string | Symbol = Symbol()) {
  return [
    function provide(v: T) {
      const vm = getCurrentInstance()
      vm?.appContext.app.provide(key, v)
    },
    function use(fallback?: T) {
      return inject(key, fallback) as T
    },
  ] as const
}

```
